### PR TITLE
search analysis nb, sql for sampling search queries over time period,…

### PR DESCRIPTION
**Pull request by ZacharyForrest**
… helper function for querying search api

# Summary

Add your summary here - keep it brief, to the point, and in plain English. For further information about pull requests,
check out the [GDS Way](https://gds-way.cloudapps.digital/standards/pull-requests.html).

# Checklists

### General checks

- [ ] Code runs
- [x] Developments are **secure** and [**ethical**](https://www.gov.uk/government/publications/data-ethics-framework))
- [x] You have made proportionate checks that the code works correctly
- [x] Test suite passes
- [x] Assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if necessary
- [ ] [Minimum usable documentation](http://agilemodeling.com/essays/documentLate.htm) written in the `docs` folder

### Project-specific checks

> These are additional checks determined necessary according to the analytical quality assurance plan
(see `docs/aqa/aqa_plan.md`).

- [ ] Check 1
- [ ] Check 2
- [ ] Check 3

### CODE REVIEW

File: notebooks/search-analysis/search_analysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "search_df['date'] = search_df['search_timestamp'].map(lambda x:x[:7])\n",
+    "search_df['tokens'] = search_df['search_term'].map(lambda x:x.split())\n",
+    "search_df['query_length'] = search_df['tokens'].map(len)"
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:106]: Interesting that this is faster than the below!

```python
    search_df['date'] = search_df['search_timestamp'].str[:7]
    search_df['tokens'] = search_df['search_term'].str.split()
    search_df['query_length'] = search_df['tokens'].str.len()
```

File: notebooks/search-analysis/search_analysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Query length over time\n",
+    "\n",
+    "import matplotlib.pyplot as plt\n",
+    "avg_len=search_df[['date','query_length']].groupby('date').mean().reset_index()\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:127]: We've already created this in Line 115. Do we need to do so again?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    "%matplotlib inline\n",
+    "plt.style.use('ggplot')\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:129]: ggplot stylings. 😎

File: notebooks/search-analysis/search_analysis.ipynb [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Query length over time\n",
+    "\n",
+    "import matplotlib.pyplot as plt\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:126]: Can this call to the matplotlib package go at the top of the notebook?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    "# Unknown: mysterious!\n",
+    "\n",
+    "\n",
+    "# annotation = annotate(df['search_term'],options=['specific_service','specific_guidance','guidance','contact','unknown'])\n",
+    "# query_cats =pd.DataFrame(annotation, columns=['query','category'])\n",
+    "query_cats = pd.read_csv(PROCESSED_DATA + '/query_categories.csv')\n",
+    "query_cat=query_cats.groupby('category').size().reset_index(name='cat_count').sort_values('cat_count', ascending=False)\n",
+    "query_cat['Percentage']=query_cat['cat_count'] / query_cat['cat_count'].sum() *100\n",
+    "import matplotlib.pyplot as plt\n",
+    "%matplotlib inline\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:182]: Already called the import of the `matplotlib` package.
```suggestion

```

File: notebooks/search-analysis/search_analysis.ipynb [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Query length hist (looks exponentially)\n",
+    "import matplotlib.pyplot as plt\n",
+    "%matplotlib inline\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:146]: Already called the import of the `matplotlib` package.
```suggestion

```

File: notebooks/search-analysis/search_analysis.ipynb [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+    "# Specific service: query names (or almost names) specific service or form (regardless of whether that service is on gov.uk)\n",
+    "# Specific guidance: query names specific guidance/stats doc on gov uk\n",
+    "# Guidance: query references topic user presumably wants guidance about\n",
+    "# Contact: query explicitly asks for contact, email, phone number, chat etc\n",
+    "# Unknown: mysterious!\n",
+    "\n",
+    "\n",
+    "# annotation = annotate(df['search_term'],options=['specific_service','specific_guidance','guidance','contact','unknown'])\n",
+    "# query_cats =pd.DataFrame(annotation, columns=['query','category'])\n",
+    "query_cats = pd.read_csv(PROCESSED_DATA + '/query_categories.csv')\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:178]: Could we call this file from within the `data/` folder in the repo instead for better reproducibility?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# Specific service: query names (or almost names) specific service or form (regardless of whether that service is on gov.uk)\n",
+    "# Specific guidance: query names specific guidance/stats doc on gov uk\n",
+    "# Guidance: query references topic user presumably wants guidance about\n",
+    "# Contact: query explicitly asks for contact, email, phone number, chat etc\n",
+    "# Unknown: mysterious!\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:173]: lol

File: notebooks/search-analysis/search_analysis.ipynb (hunk 7 above)
Comment (avisionh) [reply notebooks/search-analysis/search_analysis.ipynb:178]: Also, good to have some notes on how this file was created. I presume it was govNER?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#precision at k (k=5)\n",
+    "no_bestbets = relevancy_judgements[relevancy_judgements['best_bet']==False]\n",
+    "precision = no_bestbets[['query_length','session_id','search_term','score']][no_bestbets['score']>1].groupby(['search_term','session_id','query_length']).count()\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:257]: How was the `score` calculated?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Avg Precision at 5 \n",
+    "\n",
+    "from scipy.stats import sem, t\n",
+    "from scipy import mean\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:271]: These two package/function set calls can be placed at the top of the notebook.

File: notebooks/search-analysis/search_analysis.ipynb [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+    "avg_score = precision[['query_length','score']].groupby('query_length').agg(['mean','size','sem'])\n",
+    "avg_score = avg_score.droplevel(0,1)\n",
+    "avg_score['h'] = avg_score['sem'] * t.ppf((1 + 0.95) / 2, avg_score['size'] - 1)\n",
+    "avg_score['lower'] = avg_score['mean'] - avg_score['h']\n",
+    "avg_score['upper'] = avg_score['mean'] + avg_score['h']\n",
+    "avg_score = avg_score.reset_index()\n",
+    "\n",
+    "import matplotlib.pyplot as plt\n",
+    "%matplotlib inline\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:280]: Already called the import of the `matplotlib` package.
```suggestion

```

File: notebooks/search-analysis/search_analysis.ipynb [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "%matplotlib inline\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:322]: Already called the `matplotlib` inline magic above.
```suggestion

```

File: notebooks/search-analysis/search_analysis.ipynb [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Estimate the rate of returning a very relevant result within top 5 results\n",
+    "\n",
+    "from scipy.stats import beta\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:355]: Can this import of package go to the top of the notebook?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+    "\n",
+    "a,b = 45,36\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:357]: How did you determine the use of this numbers for the shape parameters of the beta distribution? I might need a stats refresher!

**Review by avisionh — CHANGES_REQUESTED** [review]
Thanks for this! Looks good. 🙂 I've ran it after a couple of minor tweaks to do with importing the data. A few comments but they're mainly repetitions of:
1. Can you import the data from a folder within the repo, such as `data/raw/`?
1. Can you include some additional notes on how you generated the data, where appropriate?
1. Can you move the package calls to the top of the notebook?
1. Can you remove duplicate calls to the same package?

File: notebooks/search-analysis/search_analysis.ipynb [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+    "from google.cloud import bigquery\n",
+    "import pandas_gbq\n",
+    "import nltk\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "import os\n",
+    "from pigeon import annotate\n",
+    "import json\n",
+    "from src.utils.query_search_api import query_search_api\n",
```
[the remark this answers is not in the document]
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:19]: didja source .envrc?

`# Add the working directory to PYTHONPATH; allows Jupyter notebooks in the `notebooks` folder to import `src`
export PYTHONPATH="$PYTHONPATH:$(pwd)"`

File: notebooks/search-analysis/search_analysis.ipynb (hunk 10 above)
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:271]: done-zo

File: notebooks/search-analysis/search_analysis.ipynb [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+    "client = create_big_query_client()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "best_bets = pd.read_csv(PROCESSED_DATA + '/best_bets.csv')\n",
```
[the remark this answers is not in the document]
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:65]: done!

File: notebooks/search-analysis/search_analysis.ipynb [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pd.set_option('display.max_columns', 20)\n",
+    "pd.set_option('display.max_rows',20)\n",
+    "pd.set_option('display.max_colwidth', None)\n",
+    "PROCESSED_DATA = os.getenv('DIR_DATA_PROCESSED')"
```
[the remark this answers is not in the document]
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:33]: kind've want to be able to open the file regardless of where the nb is running. What about 

`best_bets = pd.read_csv(PROJECT_DIR+'/data/processed/best_bets.csv')`

File: notebooks/search-analysis/search_analysis.ipynb (hunk 11 above)
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:280]: gone!

File: notebooks/search-analysis/search_analysis.ipynb (hunk 12 above)
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:322]: gone!

File: notebooks/search-analysis/search_analysis.ipynb (hunk 13 above)
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:355]: yep

File: notebooks/search-analysis/search_analysis.ipynb (hunk 14 above)
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:357]: not hardcoded any more!

a= number of times v relevant result returned 
b = number of times v relevant result not returned

File: notebooks/search-analysis/search_analysis.ipynb [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+    "import nltk\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import numpy as np\n",
+    "import os\n",
+    "from pigeon import annotate\n",
+    "import json\n",
+    "from src.utils.query_search_api import query_search_api\n",
+    "import re\n",
+    "import regex"
```
[the remark this answers is not in the document]
Comment (ZacharyForrest) [reply notebooks/search-analysis/search_analysis.ipynb:21]: done

File: notebooks/search-analysis/search_analysis.ipynb (hunk 17 above)
[the remark this answers is not in the document]
Comment (avisionh) [reply notebooks/search-analysis/search_analysis.ipynb]: Makes sense, happy!

File: notebooks/search-analysis/search_analysis.ipynb (hunk 15 above)
[the remark this answers is not in the document]
Comment (avisionh) [reply notebooks/search-analysis/search_analysis.ipynb:19]: Whoops, my bad! I didn't do the `direnv allow`! Done it now and should work.

File: notebooks/search-analysis/search_analysis.ipynb [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Query length over time\n",
+    "avg_len=search_df[['date','query_length']].groupby('date').mean().reset_index()\n",
+    "avg_len=search_df[['date','query_length']].groupby('date').mean().reset_index()\n",
```
Comment (avisionh) [inline notebooks/search-analysis/search_analysis.ipynb:188]: We don't need this repeated line?
```suggestion

```
